packages/io/json-deserializer/index.js

```diff
@@ -1,110 +1,100 @@
+ * @param {String} [options.version='0.0.0'] - version number to add to the metadata
+ * @param {Boolean} [options.addMetadata=true] - toggle injection of metadata at the start of the script
+ * @param {String} input - JSON source data
+ * @return {[geometry]/String} either an array of objects (geometry) or a string (script)
+ */
+const deserialize = (options, input) => {
+  const defaults = {
+    filename: 'json',
+    output: 'script',
+    version,
+    addMetaData: true
   }
-  return ''
-};
+  options = Object.assign({}, defaults, options)
+
+  // convert the JSON notation into anonymous object(s)
+  let objects = JSON.parse(input)
+
+  // cleanup the objects
+  objects = flatten(toArray(objects))
+  objects = objects.filter((object) => geometry.geom3.isA(object) || geometry.geom2.isA(object) || geometry.path2.isA(object))
+
+  return options.output === 'script' ? translate(options, objects) : objects
+}
 
 //
-// deserialize the given JSON source and return a JSCAD script
+// translate the given objects (geometries) into a  JSCAD script
 //
-// fn (optional) original filename of JSON source
+const translate = (options, objects) => {
+  const { addMetaData, filename, version } = options
+
+  let script = addMetaData ? `//
+// Produced by JSCAD IO Library : JSON Deserializer (${version})
+// date: ${new Date()}
+// source: ${filename}
 //
-function deserialize (src, fn, options) {
-  options && options.statusCallback && options.statusCallback({progress: 0})
-  fn = fn || 'amf'
-  const defaults = {version: '0.0.0'}
-  options = Object.assign({}, defaults, options)
-  const {version} = options
-
-  // convert the JSON into an anonymous object
-  var obj = JSON.parse(src)
-  options && options.statusCallback && options.statusCallback({progress: 50})
-  // convert the internal objects to JSCAD code
-  var code = ''
-  code += '//\n'
-  code += '// producer: OpenJSCAD.org ' + version + ' JSON Importer\n'
-  code += '// date: ' + (new Date()) + '\n'
-  code += '// source: ' + fn + '\n'
-  code += '//\n'
-  code += 'function main() {\n'
-  code += toSource(obj)
-  code += '};\n'
-  options && options.statusCallback && options.statusCallback({progress: 100})
-  return code
-};
+` : ''
+
+  script +=
+`
+const { geometry } = require('@jscad/modeling')
+
+const main = () => {
+  const objects = [${translateToList(objects)} ]
+  return objects
+}
+
+${translateToObjects(objects)}
+
+module.exports = { main }
+`
+
+  return script
+}
+
+const translateToList = (objects) => {
+  return objects.reduce((script, object, index) => script + ` json${index},`, '')
+}
+
+const translateToObjects = (objects) => {
+  return objects.reduce((script, object, index) => script + translateToObject(object, index), '')
+}
+
+// translate the given object to JSON notation (AGAIN)
+// NOTE: this implies that the original JSON was correct :)
+const translateToObject = (object, index) => {
+  return `const json${index} = ${JSON.stringify(object)}\n`
```

Comment:
Hmm is this not the deserializer ? ie JSON => JSCAD  ? so JSON.stringify seems wrong? should be JSON.parse so we get useable JS objects ?

Reply:
The 'object' parameter is that produced by JSON.parse(), see above. Sadly, the translation needs a string so once again JSON.stringify() is called to produce a string from the object.

Reply:
ah my bad , totally missed that, then yes, it makes sense